### PR TITLE
feat: (DHIS2-7734) add parameter to restrict associated orgunits to strictly capture scope for categoryOptions [master]

### PR DESCRIPTION
**Pull request by ameenhere**
Added a query parameter named "restrictToCaptureScope" for AbstractCrudController. It restricts the results returned to only those that have associated OrgUnits empty OR if the associated orgUnits have any capture scope ous in them. 
An additional change to queryplanner to mark the query as "nonPersisted" by recursively checking criterions for presence of non persisted property. Earlier it was only checking the first level of criterions.

### CODE REVIEW

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/query/DefaultQueryParser.java [hunk 1]
```diff
@@ -97,10 +118,34 @@ public Query parse( Class<?> klass, List<String> filters, Junction.Type rootJunc
...
 
+            if ( user != null )
+            {
+                handleCaptureScopeOuFiltering( schema, user, query.addDisjunction() );
+            }
+        }
         return query;
     }
 
+    private void handleCaptureScopeOuFiltering(Schema schema, User user, Disjunction disjunction)
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/query/DefaultQueryParser.java:134]: style, space in paranthesis

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/query/DefaultQueryService.java [hunk 2]
```diff
@@ -117,7 +117,13 @@ public Query getQueryFromUrl( Class<?> klass, List<String> filters, List<Order>
     @Override
     public Query getQueryFromUrl(Class<?> klass, List<String> filters, List<Order> orders, Pagination pagination, Junction.Type rootJunction ) throws QueryParserException
     {
-        Query query = queryParser.parse( klass, filters, rootJunction );
+        return getQueryFromUrl( klass, filters, orders, pagination, rootJunction, false );
+    }
+
+    @Override
+    public Query getQueryFromUrl(Class<?> klass, List<String> filters, List<Order> orders, Pagination pagination, Junction.Type rootJunction, boolean restrictToCaptureScope ) throws QueryParserException
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/query/DefaultQueryService.java:124]: style, space in paranthesis

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/query/QueryParser.java [hunk 3]
```diff
@@ -58,4 +60,7 @@
     Property getProperty( Schema schema, String path ) throws QueryParserException;
 
     Restriction getRestriction( Schema schema, String path, String operator, Object arg ) throws QueryParserException;
+
+    Query parse( Class<?> klass, List<String> filters, Type rootJunction, boolean restrictToCaptureScope  )
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/query/QueryParser.java:64]: Too much space in paranthesis

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/query/QueryService.java [hunk 4]
```diff
@@ -80,4 +81,7 @@
     Query getQueryFromUrl( Class<?> klass, List<String> filters, List<Order> orders, Pagination pagination) throws QueryParserException;
 
     Query getQueryFromUrl( Class<?> klass, List<String> filters, List<Order> orders ) throws QueryParserException;
+
+    Query getQueryFromUrl( Class<?> klass, List<String> filters, List<Order> orders, Pagination pagination, Type rootJunction, boolean restrictToCaptureScope  )
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/query/QueryService.java:85]: Too much space in paranthesis

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/query/DefaultQueryParser.java [hunk 5]
```diff
@@ -97,10 +118,34 @@ public Query parse( Class<?> klass, List<String> filters, Junction.Type rootJunc
                 query.add( getRestriction( schema, split[0], split[1], null ) );
             }
         }
+        
+        if ( restrictToCaptureScope && schema.haveProperty( ORGANISATION_UNITS ) )
+        {
+            User user = currentUserService.getCurrentUser();
```
[the remark this answers is not in the document]
Comment (ameenhere) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/query/DefaultQueryParser.java:124]: Not at this point. This is the part where the Query object is created. The user is yet to be injected into it. But we can add it to the Query object here then if that makes more sense?

File: dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/category/CategoryOptionController.java [hunk 6]
```diff
@@ -39,7 +50,41 @@
...
+    private CurrentUserService currentUserService;
+
+    @Autowired
+    private OrganisationUnitService organisationUnitService;
+
+    @Override
+    protected void postProcessResponseEntities( List<CategoryOption> entityList, WebOptions options, Map<String, String> parameters )
+    {
+
+        if ( "true".equalsIgnoreCase( options.getOptions().get( "restrictToCaptureScope" ) ) )
```
[the remark this answers is not in the document]
Comment (ameenhere) [reply dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/category/CategoryOptionController.java]: Isnt that happening here? if the option doesnt exist or is false, the control flow doesnt get into the if construct and jumps out. Is that OK?

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/query/DefaultQueryParser.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (mortenoh) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/query/DefaultQueryParser.java:124]: Ok, thats fine then, just leave it as it is

File: dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/category/CategoryOptionController.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (mortenoh) [reply dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/category/CategoryOptionController.java]: I mean, if you jump (return) out immediately, you don't need to nest the next lines

**Pull request merged**